drivers/soundwire/bus.c

```diff
@@ -1082,12 +1082,109 @@ int sdw_configure_dpn_intr(struct sdw_slave *slave,
 	return ret;
 }
 
+static int sdw_slave_set_frequency(struct sdw_slave *slave)
+{
+	u32 mclk_freq = slave->bus->prop.mclk_freq;
+	u32 curr_freq = slave->bus->params.curr_dr_freq >> 1;
+	u8 base_offset = 0;
+	u8 base = 0;
+	u8 scale_index = 0;
+	u32 scale;
+	int ret;
+	int i;
+
+	/*
+	 * frequency base and scale registers are required for SDCA
+	 * devices. They may also be used for 1.2+/non-SDCA devices,
+	 * but we will need a DisCo property to cover this case
+	 */
+	if (!slave->id.class_id)
+		return 0;
+
+	if (!mclk_freq) {
+		dev_err(&slave->dev,
+			"no bus MCLK, cannot set SDW_SCP_BUS_CLOCK_BASE\n");
+		return -EINVAL;
+	}
+
+	/* map base frequency using Table 89 of SoundWire 1.2 spec */
+	if (!(38400000 % mclk_freq)) {
+		base = SDW_SCP_BASE_CLOCK_19200000_HZ;
+		base_offset = 1;
+	}
+	if (!(24000000 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_24000000_HZ;
+	if (!(24576000 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_24576000_HZ;
+	if (!(22579200 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_22579200_HZ;
+	if (!(32000000 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_32000000_HZ;
```

Comment:
just to make sure: every following one is better than all previous ones, right? Because several of them might trigger. But in fact, say, if mclk_freq is a factor of both 32MHz and 24MHz, a base of 32MHz will be selected. Why is it better than a base of 24MHz? At least 24MHz should consume less?

Reply:
There is ONE mclk_freq provided by firmware. No one is trying to optimize to find the best clock, just trying to use firmware-provided information and convert into SoundWire 1.2 register values.

Reply:
In other words, this is to inform Slave the clock frequency not to set the clock frequency.

Reply:
Above code calculates the `base` variable, which is then written to the `SDW_SCP_BUS_CLOCK_BASE` register. Let's say `mclk_freq == 8000000`. The code will first enter
```
if (!(24000000 % mclk_freq))
	base = SDW_SCP_BASE_CLOCK_24000000_HZ;
```
and thus set `base` to 24MHz. But then it'll also enter
```
if (!(32000000 % mclk_freq))
	base = SDW_SCP_BASE_CLOCK_32000000_HZ;
```
and rewrite `base` to 32MHz. Is this intended? As far as I understand both 24MHz and 32MHz would work, right? Why do we then choose 32MHz? I don't know what that register does, apart from what the print below says, i.e. that it "configures the bus base," but if it does set a clock frequency to the respective value, then the lower frequency might be preferred to save power.

Reply:
no, as described in the code we don't select a clock, just mirror the settings at the Slave level.

Reply:
@plbossart Ok, so the `SDW_SCP_BUS_CLOCK_BASE` register is to inform the slave? Then in the above example it doesn't matter whether we write 24 or 32MHz there? Or is only 32MHz correct and the order is important?

Reply:
@lyakh @plbossart  Hmm I can understand this now, thanks. 

I think the if-else would be in order now that mclk_freq is set in each if branch, so the others cannot be possible be hit anymore. 

Furthermore, if the known BIOS Issue is only with 38.4Mhz, I wonder do we need the "% mclk_freq" division check at all, if this is indeed a simple mapping. I.e. could we just do:

```
switch (mclk_freq) {
 case 19200000:
        base = SDW_SCP_BASE_CLOCK_19200000_HZ;
        break;
 case 24000000:
       /* as above */
 case 38400000:
        mclk_freq = 19200000;
        base = SDW_SCP_BASE_CLOCK_19200000_HZ;
        break;
   default: 
	dev_err(&slave->dev,
			"MCLK %d is not multiple of bus curr_freq %d\n",
			mclk_freq, curr_freq);
		return -EINVAL;
 }
```

... this would be much more obvious to the reader (i.e. a simple mapping).

Reply:
> @lyakh @plbossart Hmm I can understand this now, thanks.
> 
> I think the if-else would be in order now that mclk_freq is set in each if branch, so the others cannot be possible be hit anymore.
> 
> Furthermore, if the known BIOS Issue is only with 38.4Mhz, I wonder do we need the "% mclk_freq" division check at all, if this is indeed a simple mapping. I.e. could we just do:
> 
> ```
> switch (mclk_freq) {
>  case 19200000:
>         base = SDW_SCP_BASE_CLOCK_19200000_HZ;
>         break;
>  case 24000000:
>        /* as above */
>  case 38400000:
>         mclk_freq = 19200000;
>         base = SDW_SCP_BASE_CLOCK_19200000_HZ;
>         break;
>    default: 
> 	dev_err(&slave->dev,
> 			"MCLK %d is not multiple of bus curr_freq %d\n",
> 			mclk_freq, curr_freq);
> 		return -EINVAL;
>  }
> ```
> 
> ... this would be much more obvious to the reader (i.e. a simple mapping).

I am not sure if the 38400000 case is an issue or not. But IMO we should support all valid mclk_freq. A platform (don't need to be Intel platform) may use say 12000000 mclk_freq which is not in the list but can be informed with base = 24000000 with a proper scale. So I think using % here is necessary.